File: 0167-two-sum-ii-input-array-is-sorted/167.cpp
class Solution {
public:
  vector<int> twoSum(vector<int>& numbers, int target) {
    /* Approach:
     *  1. [Two Pointer] -- T.C: O(N), S.C: (1)
     *      - The array is sorted.
     *      - Move left pointer rightwards to increase sum; and right pointer leftwards to decrease
     * sum.
     *      - Stop when sum == target.
     */

    int left  = 0;
    int right = numbers.size() - 1;

    while (left < right) {
      int sum = numbers[left] + numbers[right];
      if (sum == target)
        return {left + 1, right + 1};
      else if (sum < target)
        left++;
      else
        right--;
    }
    return {};
  }
};

File: 0167-two-sum-ii-input-array-is-sorted/167.rs
impl Solution {
    pub fn two_sum(numbers: Vec<i32>, target: i32) -> Vec<i32> {
        let (mut left, mut right): (usize, usize) = (0, numbers.len() - 1);

        while left < right {
            let sum = numbers[left] + numbers[right];
            if sum == target {
                return vec![left as i32 + 1, right as i32 + 1];
            } else if sum < target {
                left += 1;
            } else {
                right -= 1;
            }
        }
        vec![]
    }
}
